_posts/2023-07-xx-scala-as-a-junior-dev.md

```diff
@@ -0,0 +1,510 @@
+
+I see several benefits in using a functional programming style. Here are a few of them:
+
+#### 1.1.1. Predictability
+
+In functional programming, the result of a method only depends on the arguments passed to it, regardless of the state of the program. This helps to decouple the code, making it easier to test, maintain, and debug. For instance, let's say we want to add 1 to all the elements of a list:
+
+```scala
+// result depends on the argument
+var list = List(1, 2, 3)
+def add1ToAll(l: List[Int]): List[Int] = l.map(x => x + 1)
+add1ToAll(list) // List(2, 3, 4)
+list = List(2, 3, 4)
+add1ToAll(list) // List(3, 4, 5)
+```
+
+With this functional programming compliant implementation, I can predict the result without even computing it, because it depends only on the list I feed it with. I could even replace the function call with its result, and it would still work as expected.
+
+However, if I do not comply with functional programming principles, I can't be sure of the result without computing it, because it does not depend on the input anymore, but on the state of the program, which can change at any time:
+
+```scala
+// result depends on the state of the program
+var list = List(1, 2, 3)
+def add1ToAll(): List[Int] = list.map(x => x + 1)
+add1ToAll() // List(2, 3, 4)
+// list is changed somewhere else in the program
+add1ToAll() // List(3, 4, 5)
+```
+
+This produces non-deterministic code with side-effects, which can lead to bugs that are hard to find because they might appear only after a very specific and/or long sequence of events.
+
+Another benefit of the functional programming approach is that it makes it easier to reason about the code. We can have "local" reasoning, meaning that we can operate on a scope without having to worry about the rest of the program:
+
+```scala
+// non-local reasoning
+var mutableThing = ???
+def smallSpaghetti() =
+  if mutableThing == ??? then // do something
+  else // do something else
+```
+
+Here I need to understand what is `mutableThing`, when and where it is modified to understand what the method does. This is not the case in functional programming, where we can have local reasoning:
+
+```scala
+// local reasoning
+def complicateMethod(mutableThing: Any) =
+  if mutableThing == ??? then // do something
+  else // do something else
+```
+
+#### 1.1.2. Readability and reusability
+
+Being able to pass functions as arguments, and to return functions, helps to write more generic (thus reusable) code, and to focus more on the algorithm itself, rather than the implementation details. For instance, let's say we want to transform a collection. In FP, all we need to do is define a method that takes a transformation function as argument, and apply it to the collection. Then the developer can choose the transformation function they want to apply, and the method will do the rest. An example with a simple `map` method:
+
+```scala
+// FP-style
+class MyArray[A](size: Int) {
```

Comment:
I think this example implementation can confuse beginners because the implementation of map is imperative, so the next question is, "why should I use map when I can use loops - look, even you are using a loop!"

I think we can avoid the implementation here and just specify the signature and a "..." for the implementation. We can also show a link to the Scala Essentials course, because in the course I give a real example exactly with an immutable container like this, implemented functionally (not with loops).